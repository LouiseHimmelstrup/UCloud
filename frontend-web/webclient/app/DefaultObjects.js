import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};