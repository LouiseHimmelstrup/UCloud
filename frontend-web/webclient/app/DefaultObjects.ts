import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState} from "Files";
import {SortOrder, SortBy} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Publication} from "Zenodo";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum AnalysesStatusMap {
    "PENDING",
    "IN PROGRESS",
    "COMPLETED"
}

export enum RightsNameMap {
    "NONE" = "None",
    "READ" = "Read",
    "READ_WRITE" = "Read/Write"
}

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

function getFilesSortingColumnOrDefault(columnIndex: 0 | 1): SortBy {
    const sortingColumn = window.localStorage.getItem(`filesSorting${columnIndex}`);
    if (sortingColumn && Object.values(SortBy).includes(sortingColumn)) return sortingColumn as SortBy;
    switch (columnIndex) {
        case 0:
            window.localStorage.setItem("filesSorting0", SortBy.MODIFIED_AT);
            return SortBy.MODIFIED_AT;
        case 1:
            window.localStorage.setItem("filesSorting1", SortBy.SIZE);
            return SortBy.SIZE;
    }
}

function getItemOrDefault<T, T2>(itemName: string, defaultValue: T, en: T2): T {
    const item = window.localStorage.getItem(itemName);
    if (item && Object.values(en).includes(item)) {
        return item as unknown as T;
    }
    return defaultValue;
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FilesReduxObject extends ComponentWithPage<File> {
    sortOrder: SortOrder
    sortBy: SortBy
    path: string
    filesInfoPath: string
    fileSelectorError?: string
    sortingColumns: [SortBy, SortBy]
    fileSelectorLoading: boolean
    fileSelectorShown: boolean
    fileSelectorPage: Page<File>
    fileSelectorPath: string
    fileSelectorCallback: Function
    fileSelectorIsFavorites: boolean
    disallowedPaths: string[]
    invalidPath: boolean
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface ZenodoReduxObject extends ComponentWithPage<Publication> {
    connected: boolean
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    files?: Reducer<FilesReduxObject>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    zenodo?: Reducer<ZenodoReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
    detailedResult?: Reducer<DetailedResultReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

export type DetailedResultReduxObject = ComponentWithPage<File>

export const initDetailedResult = (): DetailedResultReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined
});

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    files: FilesReduxObject,
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    zenodo: ZenodoReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    detailedResult: DetailedResultReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects &
    FavoritesRedux.Objects &
    SnackbarRedux.Wrapper;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        files: initFiles(homeFolder),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        zenodo: initZenodo(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        detailedResult: initDetailedResult(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        ...FavoritesRedux.init(),
        ...SnackbarRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.ASCENDING
});

export const initZenodo = (): ZenodoReduxObject => ({
    connected: false,
    loading: false,
    page: emptyPage
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFiles = (homeFolder: string): FilesReduxObject => ({
    page: emptyPage,
    sortOrder: getItemOrDefault("sortOrder", SortOrder.ASCENDING, SortOrder),
    sortBy: getItemOrDefault("sortBy", SortBy.PATH, SortBy),
    loading: false,
    error: undefined,
    path: "",
    filesInfoPath: "",
    sortingColumns: [getFilesSortingColumnOrDefault(0), getFilesSortingColumnOrDefault(1)],
    fileSelectorLoading: false,
    fileSelectorShown: false,
    fileSelectorPage: emptyPage,
    fileSelectorPath: homeFolder,
    fileSelectorIsFavorites: false,
    fileSelectorCallback: () => undefined,
    fileSelectorError: undefined,
    disallowedPaths: [],
    invalidPath: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
