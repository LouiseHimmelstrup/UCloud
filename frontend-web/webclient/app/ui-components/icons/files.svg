<svg viewBox="0 0 24 22" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linejoin="round" stroke-miterlimit="1.414"><path d="M0 21.315c0 .378.27.688.6.688h22.8c.33 0 .6-.31.6-.688v-16.5c0-.378-.27-.687-.6-.687H10.8L7.2.003H.6c-.33 0-.6.309-.6.687v20.625z" fill="#8393a7"/><g transform="matrix(.6 0 0 .6875 0 .69)"><clipPath id="a"><path d="M0-1h32.822L8.575 31H0V-1z"/></clipPath><g clip-path="url(#a)"><path d="M0 30c0 .55.45 1 1 1h38c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1H18l-6-6H1C.45-1 0-.55 0 0v30z" fill="#53657d"/></g></g></svg>