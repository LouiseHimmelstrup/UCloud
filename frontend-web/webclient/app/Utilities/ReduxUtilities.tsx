import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    dashboard,
    header,
    status,
    sidebar,
    notifications,
    simpleSearch,
    hookStore,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
