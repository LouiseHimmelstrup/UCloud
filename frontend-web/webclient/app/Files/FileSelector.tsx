import * as React from "react";
import { Modal, Button, List, Input } from "semantic-ui-react";
import { List as PaginationList } from "Pagination/List";
import { Cloud } from "Authentication/SDUCloudObject";
import { BreadCrumbs } from "Breadcrumbs/Breadcrumbs";
import * as PropTypes from "prop-types";
import { replaceHomeFolder, getFilenameFromPath, getParentPath, isDirectory, createFolder } from "Utilities/FileUtilities";
import * as uf from "UtilityFunctions";
import PromiseKeeper from "PromiseKeeper";
import { changeUppyRunAppOpen } from "Uppy/Redux/UppyActions";
import { KeyCode } from "DefaultObjects";
import { FileIcon } from "UtilityComponents";
import "./Files.scss";
import "Styling/Shared.scss";
import { emptyPage } from "DefaultObjects";
import { FileSelectorProps, FileSelectorState, FileListProps, FileSelectorModalProps, FileSelectorBodyProps } from ".";
import { filepathQuery, isInvalidPathName } from "Utilities/FileUtilities";

class FileSelector extends React.Component<FileSelectorProps, FileSelectorState> {
    constructor(props, context) {
        super(props, context);
        this.state = {
            promises: new PromiseKeeper(),
            path: `${Cloud.homeFolder}`,
            loading: false,
            page: emptyPage,
            modalShown: false,
            uppyOnUploadSuccess: null,
            creatingFolder: false
        };
    }

    static contextTypes = {
        store: PropTypes.object.isRequired
    }

    handleKeyDown = (key, name) => {
        if (key === KeyCode.ESC) {
            this.setState(() => ({ creatingFolder: false }));
        } else if (key === KeyCode.ENTER) {
            const { path, page } = this.state;
            const fileNames = page.items.map((it) => getFilenameFromPath(it.path));
            if (isInvalidPathName(name, fileNames)) { return }
            const newPath = `${path.endsWith("/") ? path + name : path + "/" + name}`;
            createFolder(newPath, Cloud, () => {
                this.setState(() => ({ creatingFolder: false }));
                this.fetchFiles(path, page.pageNumber, page.itemsPerPage);
            });
        }
    }

    uppyOnUploadSuccess = (file, resp, uploadURL) => {
        if (!this.props.allowUpload) return;
        // TODO This is a hack.
        let apiIndex = uploadURL.indexOf("/api/");
        if (apiIndex === -1) throw "Did not expect upload URL to not contain /api/";

        let apiEndpoint = uploadURL.substring(apiIndex + 5);

        Cloud.head(apiEndpoint).then(it => {
            console.log("Got a response back!");
            let path = it.request.getResponseHeader("File-Location");
            path.lastSlash
            let lastSlash = path.lastIndexOf("/");
            if (lastSlash === -1) throw `Could not parse name of path: ${path}`;
            let name = path.substring(lastSlash + 1);
            let fileObject = {
                path: path,
                name: name,
            };
            this.props.onFileSelect(fileObject);
        });
    };

    componentDidMount() {
        const { page } = this.state;
        this.fetchFiles(Cloud.homeFolder, page.pageNumber, page.itemsPerPage);
    }

    componentWillUnmount() {
        this.state.promises.cancelPromises();
    }

    onUppyClose = (): void => {
        this.props.uppy.off("upload-success", this.state.uppyOnUploadSuccess);
        this.setState(() => ({
            uppyOnUploadSuccess: null,
        }));
    }

    setSelectedFile = (file) => {
        let fileCopy = { path: file.path };
        this.setState(() => ({
            modalShown: false,
            creatingFolder: false
        }));
        this.props.onFileSelect(fileCopy);
    }

    fetchFiles = (path: string, pageNumber: number, itemsPerPage: number) => {
        this.setState(() => ({ loading: true, creatingFolder: false }));
        this.state.promises.makeCancelable(Cloud.get(filepathQuery(path, pageNumber, itemsPerPage))).promise.then(({ response }) =>
            this.setState(() => ({
                page: response,
                loading: false,
                path
            }))
        );
    }

    render() {
        const onUpload = () => {
            if (!this.props.allowUpload) return;
            this.context.store.dispatch(changeUppyRunAppOpen(true));
            let uppy = this.props.uppy;
            uppy.reset();
            uppy.once("upload-success", this.uppyOnUploadSuccess);
        };
        const path = this.props.path ? this.props.path : "";
        const uploadButton = this.props.allowUpload ? (<UploadButton onClick={onUpload} />) : null;
        const removeButton = this.props.remove ? (<RemoveButton onClick={this.props.remove} />) : null;
        return (
            <React.StrictMode>
                <Input
                    className="readonly mobile-padding"
                    required={this.props.isRequired}
                    placeholder={"No file selected"}
                    value={path}
                    action
                >
                    <input onClick={() => this.setState(() => ({ modalShown: true }))} />
                    {uploadButton}
                    {removeButton}
                </Input>
                <FileSelectorModal
                    show={this.state.modalShown}
                    onHide={() => this.setState(() => ({ modalShown: false }))}
                    path={this.state.path}
                    navigate={this.fetchFiles}
                    page={this.state.page}
                    loading={this.state.loading}
                    creatingFolder={this.state.creatingFolder}
                    setSelectedFile={this.setSelectedFile}
                    fetchFiles={this.fetchFiles}
                    handleKeyDown={this.handleKeyDown}
                    createFolder={() => this.setState(() => ({ creatingFolder: true }))}
                    canSelectFolders={this.props.canSelectFolders}
                    onlyAllowFolders={this.props.onlyAllowFolders}
                />
            </React.StrictMode>)
    }
}

export const FileSelectorModal = (props: FileSelectorModalProps) => (
    <Modal open={props.show} onClose={props.onHide} closeOnDimmerClick size="large">
        <Modal.Header>
            File selector
            <Button circular floated="right" icon="cancel" type="button" onClick={props.onHide} />
            <Button icon="redo" loading={props.loading} floated="right" circular onClick={() => props.fetchFiles(props.path, props.page.pageNumber, props.page.itemsPerPage)} />
            <CreateFolderButton createFolder={props.createFolder} />
        </Modal.Header>
        <Modal.Content scrolling>
            <BreadCrumbs currentPath={props.path} navigate={(path) => props.fetchFiles(path, props.page.pageNumber, props.page.itemsPerPage)} />
            <PaginationList
                errorMessage={props.errorMessage}
                onErrorDismiss={props.onErrorDismiss}
                pageRenderer={(page) =>
                    <FileSelectorBody
                        {...props}
                        page={page}
                        fetchFiles={(path) => props.fetchFiles(path, page.pageNumber, page.itemsPerPage)}
                    />
                }
                page={props.page}
                onPageChanged={(pageNumber) => props.fetchFiles(props.path, pageNumber, props.page.itemsPerPage)}
                onItemsPerPageChanged={(itemsPerPage) => props.fetchFiles(props.path, 0, itemsPerPage)}
                loading={props.loading}
            />
        </Modal.Content>
    </Modal>
);

const FileSelectorBody = ({ disallowedPaths = [] as string[], onlyAllowFolders = false, ...props }: FileSelectorBodyProps) => {
    let f = onlyAllowFolders ? props.page.items.filter(f => isDirectory(f)) : props.page.items;
    const files = f.filter((it) => !disallowedPaths.some((d) => d === it.path));
    const { path } = props;
    return (
        <React.Fragment>
            <List divided size="large">
                <List.Header>
                    Filename
                </List.Header>
                <CreatingFolder
                    creatingFolder={props.creatingFolder}
                    handleKeyDown={props.handleKeyDown}
                />
                <MockFolder // Return folder
                    predicate={uf.removeTrailingSlash(path) !== uf.removeTrailingSlash(Cloud.homeFolder) && !(disallowedPaths.some(it => it === getParentPath(path)))}
                    folderName=".."
                    path={getParentPath(path)}
                    canSelectFolders={props.canSelectFolders}
                    setSelectedFile={props.setSelectedFile}
                    fetchFiles={props.fetchFiles}
                />
                <MockFolder // Current folder
                    predicate={onlyAllowFolders && !disallowedPaths.some(dP => uf.addTrailingSlash(dP) === uf.addTrailingSlash(path))}
                    path={path}
                    setSelectedFile={props.setSelectedFile}
                    canSelectFolders
                    folderName={`${getFilenameFromPath(replaceHomeFolder(path, Cloud.homeFolder))} (Current folder)`}
                    fetchFiles={() => null}
                />
                <FileList files={files} setSelectedFile={props.setSelectedFile} fetchFiles={props.fetchFiles} canSelectFolders={props.canSelectFolders} />
            </List>
        </React.Fragment>)
};

type CreateFolderButton = { createFolder: () => void }
const CreateFolderButton = ({ createFolder }: CreateFolderButton) =>
    !!createFolder ?
        (<Button onClick={() => createFolder()} basic className="float-right" content="Create new folder" />) : null;


function MockFolder({ predicate, path, folderName, fetchFiles, setSelectedFile, canSelectFolders }) {
    const folderSelection = canSelectFolders ? (
        <List.Content floated="right">
            <Button onClick={() => setSelectedFile({ path })} content="Select" />
        </List.Content>
    ) : null;
    return predicate ? (
        <List.Item className="pointer-cursor itemPadding" onClick={() => fetchFiles(path)}>
            <List.Content>
                {folderSelection}
                <List.Icon name="folder" color="blue" />
                {folderName}
            </List.Content>
        </List.Item>
    ) : null;
}

const CreatingFolder = ({ creatingFolder, handleKeyDown }) =>
    !creatingFolder ? null : (
        <List.Item className="itemPadding">
            <List.Content>
                <List.Icon name="folder" color="blue" />
                <Input
                    onKeyDown={(e) => handleKeyDown(e.keyCode, e.target.value)}
                    placeholder="Folder name..."
                    autoFocus
                    transparent
                />
                <Button floated="right" onClick={() => handleKeyDown(KeyCode.ESC)}>✗</Button>
            </List.Content>
        </List.Item>
    );

const UploadButton = ({ onClick }) => (<Button type="button" content="Upload File" onClick={onClick} />);
const RemoveButton = ({ onClick }) => (<Button type="button" content="✗" onClick={onClick} />);
const FolderSelection = ({ canSelectFolders, setSelectedFile }) => canSelectFolders ?
    (<Button onClick={setSelectedFile} floated="right" content="Select" />) : null;

const FileList = ({ files, fetchFiles, setSelectedFile, canSelectFolders }: FileListProps) =>
    !files.length ? null :
        (<React.Fragment>
            {files.map((file, index) =>
                file.type === "FILE" ? (
                    <List.Item key={index} className="itemPadding pointer-cursor">
                        <List.Content onClick={() => setSelectedFile(file)}>
                            <List.Icon name={uf.iconFromFilePath(file.path, file.type, Cloud.homeFolder)} />
                            {getFilenameFromPath(file.path)}
                        </List.Content>
                    </List.Item>
                ) : (
                        <List.Item key={index} className="itemPadding pointer-cursor">
                            <List.Content floated="right">
                                <FolderSelection canSelectFolders={canSelectFolders} setSelectedFile={() => setSelectedFile(file)} />
                            </List.Content>
                            <List.Content onClick={() => fetchFiles(file.path)}>
                                <FileIcon size={null} name="folder" link={file.link} color="blue" />
                                {getFilenameFromPath(file.path)}
                            </List.Content>
                        </List.Item>
                    ))}
        </React.Fragment>);

export default FileSelector;