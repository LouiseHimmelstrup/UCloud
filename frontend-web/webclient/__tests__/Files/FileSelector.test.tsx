import * as React from "react";
import * as Renderer from "react-test-renderer";
import FileSelector from "Files/FileSelector";
import files from "Files/Redux/FilesReducer";
import { initFiles } from "DefaultObjects";
import { configureStore } from "Utilities/ReduxUtilities";
import { Provider } from "react-redux";

const emptyPageStore = configureStore({ files: initFiles({ homeFolder: "/home/user@test.abc/" }) }, { files });

describe("File Selector", () => {
    test("Test muter", () =>
        expect(1).toBe(1)
    );

    // TODO Tests (Currently depends on the Cloud object)
});

/* :::Props:: */
// remove
// uppy

describe("File Selector Modal", () => {
    test("Minimal FileSelector", () =>
        expect(Renderer.create(
            <Provider store={emptyPageStore}>
                <FileSelector
                    onFileSelect={(f) => f}
                    path="/home/Folder/Fawlder"
                />
            </Provider>
        ).toJSON()).toMatchSnapshot());

    test("Minimal FileSelector, allow upload", () =>
        expect(Renderer.create(
            <Provider store={emptyPageStore}>
                <FileSelector
                    allowUpload
                    onFileSelect={(f) => f}
                    path="/home/Folder/Fawlder"
                />
            </Provider>
        ).toJSON()).toMatchSnapshot());

    test("Minimal FileSelector, is required", () =>
        expect(Renderer.create(
            <Provider store={emptyPageStore}>
                <FileSelector
                    isRequired
                    onFileSelect={(f) => f}
                    path="/home/Folder/Fawlder"
                />
            </Provider>
        ).toJSON()).toMatchSnapshot());

    test("Minimal FileSelector, can select folders", () =>
        expect(Renderer.create(
            <Provider store={emptyPageStore}>
                <FileSelector
                    canSelectFolders
                    onFileSelect={(f) => f}
                    path="/home/Folder/Fawlder"
                />
            </Provider>
        ).toJSON()).toMatchSnapshot());

    test("Minimal FileSelector, only allow folders", () =>
        expect(Renderer.create(
            <Provider store={emptyPageStore}>
                <FileSelector
                    onlyAllowFolders
                    onFileSelect={(f) => f}
                    path="/home/Folder/Fawlder"
                />
            </Provider>
        ).toJSON()).toMatchSnapshot());
});

