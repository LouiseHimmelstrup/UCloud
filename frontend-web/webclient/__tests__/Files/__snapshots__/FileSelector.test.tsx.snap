// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Selector Modal Minimal FileSelector 1`] = `
<div
  className="ui action input readonly mobile-padding"
>
  <input
    onChange={[Function]}
    onClick={[Function]}
    placeholder="No file selected"
    type="text"
    value="/home/Folder/Fawlder"
  />
</div>
`;

exports[`File Selector Modal Minimal FileSelector, allow upload 1`] = `
<div
  className="ui action input readonly mobile-padding"
>
  <input
    onChange={[Function]}
    onClick={[Function]}
    placeholder="No file selected"
    type="text"
    value="/home/Folder/Fawlder"
  />
  <button
    className="ui button"
    onClick={[Function]}
    role="button"
    type="button"
  >
    Upload File
  </button>
</div>
`;

exports[`File Selector Modal Minimal FileSelector, can select folders 1`] = `
<div
  className="ui action input readonly mobile-padding"
>
  <input
    onChange={[Function]}
    onClick={[Function]}
    placeholder="No file selected"
    type="text"
    value="/home/Folder/Fawlder"
  />
</div>
`;

exports[`File Selector Modal Minimal FileSelector, is required 1`] = `
<div
  className="ui action input readonly mobile-padding"
>
  <input
    onChange={[Function]}
    onClick={[Function]}
    placeholder="No file selected"
    required={true}
    type="text"
    value="/home/Folder/Fawlder"
  />
</div>
`;

exports[`File Selector Modal Minimal FileSelector, only allow folders 1`] = `
<div
  className="ui action input readonly mobile-padding"
>
  <input
    onChange={[Function]}
    onClick={[Function]}
    placeholder="No file selected"
    type="text"
    value="/home/Folder/Fawlder"
  />
</div>
`;
